### ISSUE TITLE
Realm crashes while in write transaction

### ISSUE THREAD

**Issue by RosenTomov** [realm/realm-js#3054]
## Goals & Expected Results
Be able to write data in realm.

## Actual Results
Realm crashes with the following: 

```
../src/realm/util/encrypted_file_mapping.hpp:147: [realm-core-6.0.6] Assertion failed: addr >= m_addr with (addr, m_addr) =  [true, true]

0   realm.node                          0x0000000106dc27dc _ZN5realm4utilL18terminate_internalERNSt3__118basic_stringstreamIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE + 28

1   realm.node                          0x0000000106dc2d37 _ZN5realm4util19terminate_with_infoEPKcS2_lS2_OSt16initializer_listINS0_9PrintableEE + 471

2   realm.node                          0x0000000106dad1e4 _ZN5realm4util19terminate_with_infoIJRPKvRKPvEEEvPKciS9_S9_DpOT_ + 84

3   realm.node                          0x0000000106dad093 _ZN5realm4util20EncryptedFileMapping12read_barrierEPKvmPFmPKcE + 419

4   realm.node                          0x0000000106cc8984 _ZN5realm4util26do_encryption_read_barrierEPKvmPFmPKcEPNS0_20EncryptedFileMappingE + 68

5   realm.node                          0x0000000106cc5466 _ZN5realm12ArrayKeyBaseILi1EE13init_from_refEm + 86

6   realm.node                          0x0000000106cc37ef _ZNK5realm7Cluster9init_leafENS_6ColKeyEPNS_12ArrayPayloadE + 159

7   realm.node                          0x0000000106ab7757 _ZN5realm7LinkMap11set_clusterEPKNS_7ClusterE + 471

8   realm.node                          0x0000000106d40478 _ZN5realm10ParentNode5matchERNS_8ConstObjE + 344

9   realm.node                          0x0000000106b7038c _ZNK5realm5Query8find_allERNS_14ConstTableViewEmmm + 620

10  realm.node                          0x0000000106d9b015 _ZN5realm14ConstTableView7do_syncEv + 469

11  realm.node                          0x0000000106b70621 _ZN5realm5Query8find_allEmmm + 97

12  realm.node                          0x0000000106b70b25 _ZN5realm5Query8find_allERKNS_18DescriptorOrderingE + 261

13  realm.node                          0x00000001068ff42a _ZN5realm7Results27do_evaluate_query_if_neededEb + 218

14  realm.node                          0x000000010690852d _ZNO5realm7Results8snapshotEv + 147

15  realm.node                          0x0000000106908463 _ZNKR5realm7Results8snapshotEv + 59

16  realm.node                          0x00000001068b8364 _ZN5realm2js12ResultsClassINS_4node5TypesEE8snapshotEN4Napi3EnvENS5_6ObjectERNS0_9ArgumentsIS3_EERNS0_11ReturnValueIS3_EE + 70

17  realm.node                          0x00000001068b398d _ZN5realm2js4wrapIXadL_ZNS0_12ResultsClassINS_4node5TypesEE8snapshotEN4Napi3EnvENS6_6ObjectERNS0_9ArgumentsIS4_EERNS0_11ReturnValueIS4_EEEEEENS6_5ValueERKNS6_12CallbackInfoE + 260

18  realm.node                          0x0000000106839413 _ZZN4Napi10ObjectWrapIN5realm4node13WrappedObjectINS1_2js12ResultsClassINS2_5TypesEEEEEE29InstanceMethodCallbackWrapperEP10napi_env__P20napi_callback_info__ENKUlvE_clEv + 139

19  realm.node                          0x0000000106839343 _ZN4Napi10ObjectWrapIN5realm4node13WrappedObjectINS1_2js12ResultsClassINS2_5TypesEEEEEE29InstanceMethodCallbackWrapperEP10napi_env__P20napi_callback_info__ + 39

20  node                                0x000000010005eaba _ZN6v8impl12_GLOBAL__N_123FunctionCallbackWrapper6InvokeERKN2v820FunctionCallbackInfoINS2_5ValueEEE + 122

21  node                                0x000000010024a2e8 _ZN2v88internal25FunctionCallbackArguments4CallENS0_15CallHandlerInfoE + 616

22  node                                0x00000001002498a9 _ZN2v88internal12_GLOBAL__N_119HandleApiCallHelperILb0EEENS0_11MaybeHandleINS0_6ObjectEEEPNS0_7IsolateENS0_6HandleINS0_10HeapObjectEEESA_NS8_INS0_20FunctionTemplateInfoEEENS8_IS4_EENS0_16BuiltinArgumentsE + 521

23  node                                0x0000000100249012 _ZN2v88internalL26Builtin_Impl_HandleApiCallENS0_16BuiltinArgumentsEPNS0_7IsolateE + 258

24  node                                0x00000001009ce1f9 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_BuiltinExit + 57

25  ???                                 0x000027ffeff89470 0x0 + 43980196189296!!! IMPORTANT: Please send this log and info about Realm SDK version and other relevant reproduction info to help@realm.io.Abort trap: 6
```

## Steps to Reproduce
Try to write data to a cloud realm. Same data can be processed on another realm instance, but fails in our production one.

## Version of Realm and Tooling

- Realm JS SDK Version: 6.0.2
- Node or React Native:  12.18.1
- Client OS & Version: MacOS 10.15.5, Windows 10 pro 64 bit
- Which debugger for React Native: None

**Comment by kneth** [comment]
@RosenTomov Thank you for reporting. I wonder if it is related to https://github.com/realm/realm-core/issues/3766 (hitting same assertion, and you are using encryption).

**Comment by RosenTomov** [comment]
Hello @kneth, I'm not using encryption (unless realm cloud does something by default), but I found the culprit:

Replaced a filtered query
```
.filtered(`ContactType.Id = "${utils.ContactTypeEnum.Customer.Id}"\
         && FirstName = null && LastName = null && CpId = null\
         && PhoneNumbers.Id = "${dbPhone.Id}" && IsDeleted = false`);
```

with
```
.filter(
      (f) =>
        f.ContactType.Id === utils.ContactTypeEnum.Customer.Id &&
        !f.FirstName &&
        !f.LastName &&
        !f.CpId &&
        !f.IsDeleted &&
        f.PhoneNumbers.some((s) => s.Id === dbPhone.Id)
    );
```

And now, its going through.

**Comment by kneth** [comment]
@ironage Can you take a look at the query and see why it should end in a hard crash? Moreover, can we guard against it in the query parser?

**Comment by ironage** [comment]
The stack trace is related to the memory management, it is telling that the user is not applying encryption and yet the code is trying to decrypt something. This is a problem that was fixed in core 6.0.15. The reported version here is realm-js 6.0.2 which uses core 6.0.6 so upgrading to the latest realm js release should fix this. I'm going to close this issue, but feel free to reopen if there are further issues.
